packages/phasor/src/utils/a-star.ts

```diff
@@ -32,48 +32,60 @@ export class AStarRunner {
     this._sp[2] = 0;
     this._ep[2] = 0;
     this._originalEp[2] = 0;
-    this._graph = new Graph([...points], blocks);
+    this._graph = new Graph([...points], blocks, expandBlocks);
     this._init();
   }
   private _init() {
-    this._cameFrom.set(this._sp, this._originalSp);
-    this._cameFrom.set(this._originalSp, null);
-    this._cameFrom.set(this._originalEp, this._ep);
+    this._cameFrom.set(this._sp, { from: [this._originalSp], indexs: [-1] });
+    this._cameFrom.set(this._originalSp, { from: [], indexs: [] });
 
-    this._costSoFar.set(this._sp, 0);
-    this._diagonalCount.set(this._sp, 0);
-    this._pointPriority.set(this._sp, 0);
+    this._costSoFar.set(this._sp, [0]);
+    this._diagonalCount.set(this._sp, [0]);
+    this._pointPriority.set(this._sp, [0]);
     this._frontier = new PriorityQueue<
       IVec,
       [diagonalCount: number, pointPriority: number, distCost: number]
-    >((a, b) => {
-      if (a[2] + 0.01 < b[2]) return -1;
-      else if (a[2] - 0.01 > b[2]) return 1;
-      else if (a[0] < b[0]) return -1;
-      else if (a[0] > b[0]) return 1;
-      else if (a[1] > b[1]) return -1;
-      else if (a[1] < b[1]) return 1;
-      else return 0;
-    });
+    >(this._compare);
     this._frontier.enqueue(this._sp, [0, 0, 0]);
   }
 
+  private _compare(a: [number, number, number], b: [number, number, number]) {
+    if (a[2] + 0.01 < b[2]) return -1;
+    else if (a[2] - 0.01 > b[2]) return 1;
+    else if (a[0] < b[0]) return -1;
+    else if (a[0] > b[0]) return 1;
+    else if (a[1] > b[1]) return -1;
+    else if (a[1] < b[1]) return 1;
+    else return 0;
+  }
+
   private _heuristic(a: IVec, b: IVec): number {
     return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
   }
 
-  private _getDiagonalCount(a: IVec, last?: IVec, last2?: IVec): number {
-    last = last ?? this._cameFrom.get(a) ?? undefined;
-    if (!last) return 0;
-    last2 = last2 ?? this._cameFrom.get(last) ?? undefined;
-    if (!last2) return 0;
+  private _getDiagonalCount(a: IVec, last: IVec, last2: IVec): number {
     if (almostEqual(a[0], last[0]) && almostEqual(a[0], last2[0])) return 0;
     if (almostEqual(a[1], last[1]) && almostEqual(a[1], last2[1])) return 0;
     return 1;
   }
 
   private _almostEqual(a: IVec, b: IVec): boolean {
-    return almostEqual(a[0], b[0]) && almostEqual(a[1], b[1]);
+    return almostEqual(a[0], b[0], 0.02) && almostEqual(a[1], b[1], 0.02);
+  }
+
+  private _neighbors(cur: IVec) {
```

Comment:
These non-`this` dependent methods should be top-level utils, including `_compare`, `_heuristic` and more. Please try making less inconsistencies that has been addressed in former review.

Reply:
ok，I'll put more attention on it